### PR TITLE
ARROW-5037: [Rust] [DataFusion] Refactor aggregate functions

### PR DESCRIPTION
**Pull request by andygrove**
This PR splits out the various aggregate functions into separate files and adds a new `aggregate_array` method to the aggregate function trait, allowing more logic to be delegated to each function.

This PR builds on https://github.com/apache/arrow/pull/4036 and I will rebase once that one is merged.

### CODE REVIEW

File: rust/datafusion/src/execution/aggregate/count.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        "count"
+    }
+
+    fn accumulate_scalar(
+        &mut self,
+        value: &Option<ScalarValue>,
+        rollup: bool,
+    ) -> Result<()> {
+        if rollup {
+            if let Some(ScalarValue::UInt64(n)) = value {
```
Comment (bkietz) [inline rust/datafusion/src/execution/aggregate/count.rs]:
```suggestion
            self.value = match (value, self.value) {
                (Some(m), Some(n)) => Some(m + n),
                (None, Some(n)) => Some(n),
                _ => None
            };
```

File: rust/datafusion/src/execution/aggregate/max.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+use crate::logicalplan::ScalarValue;
+
+/// Implementation of MAX aggregate function
+#[derive(Debug)]
+pub(super) struct MaxFunction {
+    data_type: DataType,
+    value: Option<ScalarValue>,
+}
+
+impl MaxFunction {
```
Comment (sunchao) [inline rust/datafusion/src/execution/aggregate/max.rs]: Seems min, max and count are very similar. Is it possible to combine them into one, or reuse some of the common code?

Comment (kszucs) [reply rust/datafusion/src/execution/aggregate/max.rs]: Perhaps a trait which implements the common functionality?

**Comment by andygrove** [comment]
I'm closing this and will tackle this again another time. I am now working on parallel query execution (single process, multiple threads) and this is going to involve some refactoring of the codebase.

**Pull request closed without merging**